File: Cpp/817_Linked_List_Components.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    int numComponents(ListNode* head, vector<int>& G) {
        int count = 0;
        unordered_set<int> s(G.begin(), G.end());
        ListNode* node = head;
        bool include = false;
        while(node){
            auto found = s.find(node->val);
            if(!include && found != s.end()) ++count;
            include = found != s.end();
            node = node->next;
        }
        return count;
    }
};

File: Kotlin/817_Linked_List_Components.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun numComponents(head: ListNode?, nums: IntArray): Int {      
        var result = 0
        val s = nums.toMutableSet()
        var node = head
        var prev = false

        while (node != null) {
            val found = s.contains(node.`val`)
            if(!prev && found) ++result
            prev = found
            node = node.next
        }
        
        return result
    }
}
